### PR TITLE
Add local module with explicit imports detection

### PR DESCRIPTION
**Pull request by mbj**

### CODE REVIEW

File: src/SourceConstraints.hs [hunk 1]
```diff
@@ -48,18 +53,24 @@ runSourceConstraints :: [CommandLineOption]
...
 
+  localModules <- mapM parseLocalModule (pack <$> options)
+
   when (allowLocation ml_hs_file) $
     liftIO
       . printOrThrowWarnings dynFlags
       $ warnings Context{..} (hpm_module parsedModule)
 
   pure parsedModule
+
```
Comment (snusnu) [inline src/SourceConstraints.hs]: @mbj To be consistent with other `where` clauses, this blank line should be removed.

File: src/SourceConstraints.hs [hunk 2]
```diff
@@ -88,13 +99,48 @@ data IEClass = Module String | Type String | Operator String | Function String
...
+      , mkQ empty sortedMultipleDeriving node
+      ]
+
+    absentImportDeclList :: HsModule GhcPs -> WarningMessages
+    absentImportDeclList HsModule{..} =
+      listToBag $ catMaybes (absentList <$> candidates)
+      where
+        absentList :: LImportDecl GhcPs -> Maybe ErrMsg
+        absentList = \case
+          (L _loc ImportDecl { ideclHiding = Just (False, L loc list) }) ->
```
Comment (snusnu) [inline src/SourceConstraints.hs]: @mbj Whitespace before and after `{` (and `}`) is inconsistent with other places this construct gets used.

Comment (mbj) [reply src/SourceConstraints.hs]: I think its canonical to have that whitespace. The other places (outside the diff) are at fault.

A future pass will fix this.

The only exception is `{..}`.

File: src/SourceConstraints.hs [hunk 3]
```diff
@@ -59,7 +58,7 @@ runSourceConstraints _options ModSummary{ms_location = ModLocation{..}} parsedMo
 
   pure parsedModule
   where
-    allowLocation = maybe False (not . elem ".stack-work/" . splitPath)
+    allowLocation = maybe False (notElem ".stack-work/" . splitPath)
```
Comment (snusnu) [inline src/SourceConstraints.hs]: @mbj I think this hunk should be extracted to a prefixed `Refactor`.

Comment (mbj) [reply src/SourceConstraints.hs]: No, because its only possible to do this AFTER we brought `notElem` into scope, via removing the explicit imports.

Comment (snusnu) [reply src/SourceConstraints.hs]: Well, I am aware of that. I was suggesting to first import it and make the refactoring, then have that commit be an atomic switch to no explicit import lists.

Comment (mbj) [reply src/SourceConstraints.hs]: This is intermediary diff for no reason.

File: src/SourceConstraints.hs (hunk 1 above)
Comment (snusnu) [reply src/SourceConstraints.hs]: Bump

File: src/SourceConstraints.hs [hunk 4]
```diff
@@ -48,18 +53,24 @@ runSourceConstraints :: [CommandLineOption]
                      -> ModSummary
                      -> HsParsedModule
                      -> Hsc HsParsedModule
-runSourceConstraints _options ModSummary{ms_location = ModLocation{..}} parsedModule = do
+runSourceConstraints options ModSummary{ms_location = ModLocation{..}} parsedModule = do
   dynFlags <- getDynFlags
 
+  localModules <- mapM parseLocalModule (pack <$> options)
+
```
Comment (snusnu) [inline src/SourceConstraints.hs]: @mbj Why is this blank line added here? And why is line 58 blank too?

File: src/SourceConstraints/LocalModule.hs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+newtype LocalModule = LocalModule ModuleName
+
+localModuleParser :: Parser LocalModule
+localModuleParser = LocalModule <$> (string "local:" *> moduleNameParser)
+
+moduleNameParser :: Parser ModuleName
+moduleNameParser = do
+  first  <- section
+  others <- many (char '.' *> section)
+
```
Comment (snusnu) [inline src/SourceConstraints/LocalModule.hs]: @mbj This blank line and the one on line 29 seem redundant to me.

**Pull request merged**